### PR TITLE
validate: improve ValidateHTTPHeaderValue

### PR DESCRIPTION
**Pull request by zirain**
**Please provide a description of this PR:**

fixes: #42749

### CODE REVIEW

File: pkg/config/validation/validation.go [hunk 1]
```diff
@@ -372,16 +372,26 @@ func ValidateHTTPHeaderOperationName(name string) error {
...
 func ValidateHTTPHeaderValue(value string) error {
-	if strings.Count(value, "%")%2 != 0 {
-		return errors.New("single % not allowed.  Escape by doubling to %% or encase Envoy variable name in pair of %")
+	if !httpHeaderValueRegexp.MatchString(value) {
+		return fmt.Errorf("header value configuration %s is invalid", value)
 	}
-	return nil
+
+	if err := validateHeaderValue(value); err != nil {
+		return fmt.Errorf("header value configuration %s is invalid", value)
```
Comment (ymesika) [inline pkg/config/validation/validation.go:389]: If the validation function has the exact description of the error why not use it and provide it to the users?

File: pkg/config/validation/validation.go [hunk 2]
```diff
@@ -372,16 +372,26 @@ func ValidateHTTPHeaderOperationName(name string) error {
...
+	}
+
+	// TODO: find a better way to validate fileds supported in custom header, e.g %ENVIRONMENT(X):Z%
+
+	return validateHeaderValue(value)
```
Comment (ymesika) [inline pkg/config/validation/validation.go:394]: Why twice?

Comment (zirain) [reply pkg/config/validation/validation.go]: mistake, reverted

File: pkg/config/validation/validation.go (hunk 1 above)
Comment (zirain) [reply pkg/config/validation/validation.go]: updated

**Review by ymesika — APPROVED** [review]

File: pkg/config/validation/validation.go [hunk 3]
```diff
@@ -372,15 +372,25 @@ func ValidateHTTPHeaderOperationName(name string) error {
...
 // ValidateHTTPHeaderValue validates a header value for Envoy
 // Valid: "foo", "%HOSTNAME%", "100%%", "prefix %HOSTNAME% suffix"
-// Invalid: "abc%123"
+// Invalid: "abc%123", "%START_TIME%%"
 // We don't try to check that what is inside the %% is one of Envoy recognized values, we just prevent invalid config.
 // See: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers.html#custom-request-response-headers
 func ValidateHTTPHeaderValue(value string) error {
-	if strings.Count(value, "%")%2 != 0 {
-		return errors.New("single % not allowed.  Escape by doubling to %% or encase Envoy variable name in pair of %")
+	if !httpHeaderValueRegexp.MatchString(value) {
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:384]: Not sure I understand this. Its checking if its valid based on HTTP spec, but does it need to be? Envoy substitutes it, so it doesn't need to comply to HTTP it needs to comply to Envoy

File: pkg/config/validation/header_value_validator.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+	ExpectStringParserState                                 // expect starting " in array of strings
+	StringParserState                                       // consuming an array element string
+	ExpectArrayDelimiterOrEndParserState                    // expect array delimiter (,) or end of array (])
+	ExpectArgsEndParserState                                // expect closing ) in %VAR(...)%
+	ExpectVariableEndParserState                            // expect closing % in %VAR(...)%
+)
+
+// validateHeaderValue is golang port version of
+// https://github.com/envoyproxy/envoy/blob/master/source/common/router/header_parser.cc#L73
+func validateHeaderValue(headerValue string) error {
```
Comment (howardjohn) [inline pkg/config/validation/header_value_validator.go:37]: can we add a fuzzer? Any parser should have fuzzing. See pkg/fuzz/README.md

File: pkg/config/validation/validation.go (hunk 3 above)
Comment (zirain) [reply pkg/config/validation/validation.go:384]: envoy has proto validator [here](https://github.com/envoyproxy/envoy/blob/8674f661b216ee81ab0b2c868e09830bbe9c7df9/api/envoy/api/v2/core/base.proto#L268)

File: pkg/config/validation/header_value_validator.go (hunk 4 above)
Comment (zirain) [reply pkg/config/validation/header_value_validator.go:37]: sure, will add fuzz test later.

File: pkg/config/validation/validation.go (hunk 3 above)
Comment (kyessenov) [reply pkg/config/validation/validation.go:384]: CC @costinm 
We never claimed to support Envoy format strings, did we? I know users rely on it, but they are not subject to strict backwards compatibility guarantees of xDS. I think we may choose to only allow actual values, but that would be a breaking change.

Comment (kyessenov) [reply pkg/config/validation/validation.go:384]: For an example of upgrade issue: https://github.com/envoyproxy/envoy/issues/23972

Comment (zirain) [reply pkg/config/validation/validation.go:384]: IMO, istio's webhook should not block a valid configuration like  `'%DOWNSTREAM_PEER_CERT_V_END(%b %d %H:%M:%S %Y %Z)%'`

Comment (zirain) [reply pkg/config/validation/validation.go:384]: kindly ping @howardjohn

**Review by howardjohn — APPROVED** [review]

**Pull request merged**